cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaRoller.java

```diff
@@ -200,10 +201,17 @@ public Future<Void> rollingRestart(Function<Pod, RestartReasons> podNeedsRestart
             List<PodRef> pods = new ArrayList<>(podList.size());
 
             for (int podIndex = 0; podIndex < podList.size(); podIndex++) {
+                // Precondition: We expect the STS or SPS controller to have created the pod
+                var pod = podOperations.get(namespace, podList.get(podIndex));
+                if (pod == null) {
+                    result.fail(new KafkaRollerPreconditionFailedException("Pod doesn't exist. There seems to be some problem with the creation of pod by StatefulSets/StrimziPodSets controller"));
```

Comment:
So basically this is a pre-condition check which avoids any non nexistent pod to be reconfigured/rolled. If the pod comes out to be empty then in the next reconciliation the Kafka CR status will be updated with the error condition and we will not do any sort of rolling in this scenario . If we don't want to do have an issue when pod doesn't exist then I can probably remove the generation of exception and only stop the rolling process?

Reply:
@ShubhamRwt the question is whether aborting the whole reconciliation is the right thing to do in this edge case where a pod doesn't exist. There could be other pods in the cluster which need to be rolled, which suggests it would be better to do as Jakub suggests and for the roller to just ignore non-existent pods. That would be better than the operator being in a stuck state, failing every reconciliation with this exception until the condition is satisfied.

I think a "ignore non-existent pods" behaviour _might_ also address #4869 where the `Kafka` CR gets deleted (and GC deletes the STS/SPS and Pods), but the roller end up taking a long time to finish because the pods have disappeared under its feet. For that to work we'd need to by doing the existence check at the point where the roller was trying to interact with the pod, rather than filtering the list of pods for existence at the start of the rolling process.

Reply:
@scholzj @tombentley As per the new changes, The NPE is now removed. But yeah now instead of failing the reconciliation we are ignoring the non existent pod. I have tested the change with both podsets and statefulsets